assets/src/components/detours/detourMap.tsx

```diff
@@ -1,33 +1,56 @@
-import React, { useState } from "react"
-import { Shape } from "../../schedule"
-import { LatLngExpression } from "leaflet"
+import React, { useEffect, useState } from "react"
+import { Shape, ShapePoint } from "../../schedule"
+import { LatLngLiteral } from "leaflet"
 import { Polyline, useMap, useMapEvent } from "react-leaflet"
 import Leaflet, { Map as LeafletMap } from "leaflet"
 import Map from "../map"
 import { CustomControl } from "../map/controls/customControl"
 import { Button } from "react-bootstrap"
 import { ReactMarker } from "../map/utilities/reactMarker"
+import { fetchDetourDirections } from "../../api"
 
 export const DetourMap = ({ shape }: { shape: Shape }) => {
-  const [startPoint, setStartPoint] = useState<LatLngExpression | null>(null)
-  const [endPoint, setEndPoint] = useState<LatLngExpression | null>(null)
-  const [detourPositions, setDetourPositions] = useState<LatLngExpression[]>([])
+  const [startPoint, setStartPoint] = useState<LatLngLiteral | null>(null)
+  const [endPoint, setEndPoint] = useState<LatLngLiteral | null>(null)
+  const [waypoints, setWaypoints] = useState<LatLngLiteral[]>([])
+  const [detourShapePositions, setDetourShapePositions] = useState<
+    LatLngLiteral[]
+  >([])
 
-  const onAddDetourPosition = (p: LatLngExpression) =>
-    setDetourPositions((positions) => [...positions, p])
+  const onAddDetourPosition = (p: LatLngLiteral) => {
+    setWaypoints((positions) => [...positions, p])
+  }
+
+  useEffect(() => {
+    const shapePoints: ShapePoint[] = waypoints.map(
+      (waypoint: LatLngLiteral) => {
+        const { lat, lng } = waypoint
+        return { lat, lon: lng }
+      }
+    )
+
+    fetchDetourDirections(shapePoints).then((detourShape) => {
+      if (detourShape) {
+        setDetourShapePositions(
+          detourShape.coordinates.map((position: ShapePoint) => {
+            const { lat, lon } = position
+            return { lat, lng: lon }
+          })
+        )
+      }
+    })
+  }, [waypoints])
```

Comment:
**issue:** don't forget to handle the race condition where two `fetchDetourDirections` calls happen in quick succession and then resolve out-of-order. The `useNearestIntersection` hook contains a good example of how to guard against this.

Reply:
Ah yeah I was wondering how I should handle that. 

Just to clarify: Your suggestion here is to make the `waypoints` array part of a bigger struct that includes `loading` state as well, and to only call `fetchDetourDirections` when `loading` is `false`? NOT to create a hook.

Reply:
I'm agnostic on how to solve it, what you described is one approach but `useNearestIntersection` just uses a local variable within the `useEffect` function. As for making this its own hook I could go either way, but if the logic gets more complex it might be worth moving it out.